src/parquet/column_reader.cc

```diff
@@ -269,4 +167,36 @@ bool ColumnReader::ReadNewPage() {
   return true;
 }
 
+std::shared_ptr<ColumnReader> make_column_reader(const parquet::ColumnMetaData* metadata,
+    const parquet::SchemaElement* element, InputStream* stream) {
+  ColumnReader* result = nullptr;
+
+  switch (metadata->type) {
+    case Type::BOOLEAN:
+      result = new BoolReader(metadata, element, stream);
+      break;
+    case Type::INT32:
+      result = new Int32Reader(metadata, element, stream);
+      break;
+    case Type::INT64:
+      result = new Int64Reader(metadata, element, stream);
+      break;
+    case Type::FLOAT:
+      result = new FloatReader(metadata, element, stream);
+      break;
+    case Type::DOUBLE:
+      result = new DoubleReader(metadata, element, stream);
+      break;
+    case Type::BYTE_ARRAY:
+      result = new ByteArrayReader(metadata, element, stream);
+      break;
+    case Type::INT96:
+      result = new Int96Reader(metadata, element, stream);
+      break;
+    default:
+      ParquetException::NYI("type reader not implemented");
+  }
+  return std::shared_ptr<ColumnReader>(result);
```

Comment:
make_shared is generally preferred but probably not a big deal here

http://en.cppreference.com/w/cpp/memory/shared_ptr/make_shared